File: backend/app/inference/spatial/distance.py
"""
Distance Estimation Module.
Maps depth samples to calibrated metric distance and 5 standard distance categories:
VERY_NEAR, NEAR, MEDIUM, FAR, VERY_FAR.
"""

from typing import Tuple, Optional
from app.inference.config import DistanceConfig, pipeline_config

class DistanceEstimator:
    """Estimates distance in meters and standard navigation distance categories."""
    def __init__(self, config: DistanceConfig = None):
        self.config = config or pipeline_config.distance

    def categorize_metric(self, distance_m: float) -> str:
        """Categorizes metric distance into standard tiers."""
        cfg = self.config
        if distance_m < cfg.very_near_max_m:
            return "VERY_NEAR"
        elif distance_m < cfg.near_max_m:
            return "NEAR"
        elif distance_m < cfg.medium_max_m:
            return "MEDIUM"
        elif distance_m <= cfg.far_max_m:
            return "FAR"
        else:
            return "VERY_FAR"

    def estimate_from_depth(
        self,
        raw_depth_value: Optional[float],
        is_metric: bool = True
    ) -> Tuple[Optional[float], str]:
        """
        Estimates distance in meters and distance category.
        Returns:
            distance_m: float or None
            distance_category: "VERY_NEAR" | "NEAR" | "MEDIUM" | "FAR" | "VERY_FAR"
        """
        if raw_depth_value is None or raw_depth_value <= 0:
            return None, "FAR"

        if is_metric:
            distance_m = round(float(raw_depth_value), 2)
            category = self.categorize_metric(distance_m)
            return distance_m, category
        else:
            # Relative depth (0.0 to 1.0 or inverted)
            # Normalize to 0-10m scale for category assignment
            rel = max(0.01, min(10.0, float(raw_depth_value)))
            category = self.categorize_metric(rel)
            return round(rel, 2), category

    def estimate_from_bbox_ratio(self, area_ratio: float) -> Tuple[float, str]:
        """
        Fallback heuristic when depth is completely unavailable.
        Uses bounding box to frame area ratio.
        """
        cfg = self.config
        if area_ratio >= cfg.bbox_ratio_very_near:
            return 0.8, "VERY_NEAR"
        elif area_ratio >= cfg.bbox_ratio_near:
            return 1.5, "NEAR"
        elif area_ratio >= cfg.bbox_ratio_medium:
            return 3.0, "MEDIUM"
        elif area_ratio >= cfg.bbox_ratio_far:
            return 7.0, "FAR"
        else:
            return 12.0, "VERY_FAR"

File: frontend/src/core/distance.js
/**
 * Distance Estimator Heuristic:
 * Fallback estimation using bounding-box area ratio when dense depth is unavailable.
 */

export const DISTANCE_TIERS = {
  VERY_CLOSE: { minRatio: 0.35, label: "<1 m (~0.8m)", distanceM: 0.8, category: "VERY_NEAR", priority: "CRITICAL" },
  NEAR:       { minRatio: 0.20, label: "1–2 m (~1.5m)", distanceM: 1.5, category: "NEAR", priority: "HIGH" },
  MID:        { minRatio: 0.10, label: "3–5 m (~3.5m)", distanceM: 3.5, category: "MEDIUM", priority: "MEDIUM" },
  FAR:        { minRatio: 0.05, label: "6–10 m (~7.5m)", distanceM: 7.5, category: "FAR", priority: "LOW" },
  DISTANT:    { minRatio: 0.00, label: ">10 m", distanceM: 12.0, category: "VERY_FAR", priority: "IGNORE" }
};

export class DistanceEstimator {
  estimate(box, frameWidth, frameHeight) {
    const [x1, y1, x2, y2] = box;
    const boxWidth = Math.max(0, x2 - x1);
    const boxHeight = Math.max(0, y2 - y1);
    const boxArea = boxWidth * boxHeight;
    const frameArea = frameWidth * frameHeight;

    if (frameArea <= 0) return { ...DISTANCE_TIERS.DISTANT, areaRatio: 0 };

    const areaRatio = boxArea / frameArea;

    if (areaRatio > 0.35) return { ...DISTANCE_TIERS.VERY_CLOSE, areaRatio };
    if (areaRatio > 0.20) return { ...DISTANCE_TIERS.NEAR, areaRatio };
    if (areaRatio > 0.10) return { ...DISTANCE_TIERS.MID, areaRatio };
    if (areaRatio > 0.05) return { ...DISTANCE_TIERS.FAR, areaRatio };
    return { ...DISTANCE_TIERS.DISTANT, areaRatio };
  }
}

export const distanceEstimator = new DistanceEstimator();
